### ISSUE TITLE
Enhancing Surya with Autoregressive Generation for Extended Solar Forecasting

### ISSUE THREAD

**Issue by MartialTerran** [NASA-IMPACT/Surya#30]
### **New Issue: Enhancing Surya with Autoregressive Generation for Extended Solar Forecasting**

**Opened by:** Martial Terran mrferran1970@gmail.com https://huggingface.co/MartialTerran
**Date:** 2025-09-11

**Labels:** `enhancement`, `autoregressive`, `forecasting`

### **1. Summary**

This proposal outlines a plan to enhance the Surya foundation model with a robust autoregressive generation capability. The goal is to extend Surya's predictive horizon significantly beyond the current fixed interval, transforming it into a dynamic, long-range forecasting engine for solar phenomena. This will be achieved by creating a generative feedback loop where the model's own predicted output at time `t+1` is used as input to generate a prediction for `t+2`, and so on. This approach is inspired by recent advancements in real-time, sequential generation seen in models like Decart's "Oasis".

### **2. Background and Motivation**

The Surya model has demonstrated state-of-the-art performance in several heliophysics tasks, including solar flare forecasting. Currently, the model provides forecasts for a fixed, relatively short-term interval. However, many critical applications, such as satellite operations and space weather prediction, would benefit from longer-range forecasts (e.g., 12, 24, or even 48 hours).

By implementing an autoregressive process, we can "unroll" Surya's predictions over an extended period. This would allow us to forecast the evolution of solar features, such as active regions, and predict the likelihood of solar flares over much longer time horizons. This enhancement would represent a significant step forward in our ability to anticipate and prepare for space weather events.

### **3. Proposed Engineering Plan**

We propose a multi-phase plan to implement and evaluate the autoregressive enhancement to Surya.

**Phase 1: Baseline Autoregressive Implementation**

1.  **Develop a "rollout" script:** This script will be the core of the autoregressive process. It will:
    *   Take an initial set of SDO observations as input.
    *   Use the existing pre-trained Surya model to generate a prediction for the next time step.
    *   Feed the generated prediction back into the model as input for the subsequent prediction.
    *   Repeat this process for a specified number of steps (e.g., to generate a 24-hour forecast).

2.  **Initial Evaluation:** We will perform an initial evaluation of the baseline autoregressive implementation to assess its stability and performance. Key metrics will include:
    *   **Prediction Quality vs. Time Horizon:** How quickly does the prediction quality degrade as we extend the forecast?
    *   **Error Accumulation:** We will analyze the accumulation of errors in the generated images over time.

**Phase 2: Model Fine-Tuning for Autoregressive Stability**

Based on the findings from Phase 1, we will fine-tune the Surya model to improve its performance in an autoregressive setting. This will involve:

1.  **History Augmentation:** Inspired by Decart's approach, we will experiment with "history augmentation" during fine-tuning. This involves intentionally introducing small errors or noise into the input data during training to make the model more robust to the imperfections that will inevitably arise in a multi-step autoregressive rollout.

2.  **Loss Function Optimization:** We will investigate alternative loss functions that may be better suited for long-range forecasting. For example, we could experiment with loss functions that more heavily penalize divergences from the ground truth over longer time horizons.

**Phase 3: Integration with Downstream Tasks**

Once we have a stable and reliable autoregressive model, we will integrate it with the existing downstream tasks, particularly solar flare forecasting.

1.  **Long-Range Flare Prediction:** We will modify the solar flare forecasting task to accept the autoregressively generated image sequences as input. This will enable us to generate flare predictions for extended time horizons.

2.  **Uncertainty Quantification:** We will explore methods for quantifying the uncertainty in our long-range forecasts. This could involve techniques such as using Monte Carlo dropouts to generate an ensemble of possible future scenarios.

### **4. Code Snippets and Implementation Details**

The following pseudo-code illustrates the core logic of the proposed autoregressive rollout script:

```python
import torch
from surya.model import SuryaModel

def autoregressive_forecast(model, initial_frames, num_steps):
    """
    Generates a sequence of solar forecasts using an autoregressive process.

    Args:
        model (SuryaModel): The pre-trained or fine-tuned Surya model.
        initial_frames (torch.Tensor): The initial set of SDO observations.
        num_steps (int): The number of future time steps to predict.

    Returns:
        list[torch.Tensor]: A list of predicted future frames.
    """
    predicted_frames = []
    current_frames = initial_frames

    for _ in range(num_steps):
        # Generate the next frame prediction
        with torch.no_grad():
            next_frame = model(current_frames)

        # Add the prediction to our list
        predicted_frames.append(next_frame)

        # Update the input for the next iteration
        # Note: The exact mechanism for updating `current_frames` will
        # depend on the model's input requirements.
        current_frames = torch.roll(current_frames, shifts=-1, dims=1)
        current_frames[:, -1, ...] = next_frame

    return predicted_frames

# Example Usage
# model = SuryaModel.from_pretrained("nasa-ibm-ai4science/Surya-1.0")
# initial_data = load_sdo_data("2025-01-01T00:00:00")
# forecast = autoregressive_forecast(model, initial_data, num_steps=120) # 24-hour forecast at 12-min cadence```

### **5. Evaluation and Success Metrics**

The success of this project will be measured by the following metrics:

*   **Extended Forecast Accuracy:** We will evaluate the accuracy of our long-range forecasts against ground truth SDO data.
*   **Solar Flare Prediction Performance:** We will assess the performance of the solar flare prediction task using the extended forecasts.
*   **Model Stability:** The ability of the model to generate coherent and physically plausible solar images over long time horizons.

### **6. Conclusion**

By implementing an autoregressive generation capability, we can significantly enhance the predictive power of the Surya model. This will enable us to move from short-term forecasting to long-range prediction, providing a valuable new tool for space weather research and operations. 

Martial Terran 
mrferran1970@gmail.com 
https://huggingface.co/MartialTerran
